### ISSUE TITLE
no-invalid-this gets triggered when building object with Object.defineProperties()

### ISSUE THREAD

**Issue by ThaJay** [eslint/eslint#10187]
**Tell us about your environment**

* **ESLint Version:** v4.19.1
* **Node Version:** v9.10.1
* **npm Version:** 5.8.0

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
const error = "error"
const never = "never"
const always = "always"
const off = "off"
const warn = "warn"

module.exports = {
  "parser": "babel-eslint",
  "extends": ["standard", "standard-react"],
  "plugins": [
    "babel",
    "react",
    "jsx-a11y",
    "import"
  ],
  "globals": {
    "React"         : true,
    "__DEV__"       : true,
    "WebSocket"     : true,
    "XMLHttpRequest": true,
    "jest"          : true,
    "describe"      : true,
    "beforeAll"     : true,
    "afterAll"      : true,
    "beforeEach"    : true,
    "afterEach"     : true,
    "test"          : true,
    "expect"        : true
  },
  "rules": {
    "key-spacing"                : [warn, {
      "singleLine": {
        "beforeColon": false,
        "afterColon" : false
      },
      "align"     : {
        "beforeColon": false,
        "afterColon" : true,
        "on"         : "colon"
      }
    }],
    "new-cap"                    : off,
    "no-invalid-this"            : off,
    "object-curly-spacing"       : off,
    "quotes"                     : off,
    "semi"                       : off,
    "babel/new-cap"              : 1,
    "babel/no-invalid-this"      : 1,
    "babel/object-curly-spacing" : 1,
    "babel/quotes"               : [error, "single"],
    "babel/semi"                 : [error, never],
    "space-before-function-paren": [error, always],
    "object-curly-spacing"       : [error, never],
    "block-spacing"              : [error, never],
    "curly"                     : off, // [error, "multi-or-nest"],
    "spaced-comment"            : off,
    "space-in-parens"           : off,
    "indent"                    : off,
    "no-multi-spaces"           : off,
    "no-multiple-empty-lines"   : off,
    "padded-blocks"             : off,
    "react/prop-types"          : off,
    "react/jsx-equals-spacing"  : off,
    "import/extensions"         : off
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
export function createGetter (value) {
  return {
    'get'       : value,
    configurable: true
  }
}

export default function createModelInstance (instance) {
  return Object.defineProperties(
    {...instance},
    {
      checked: createGetter(
        function () {
          return (
            (!this.nextStatus && this.status === 'used') ||
            (this.nextStatus === 'used')
          )
        }
      )
    }
  )
}


```

```bash
eslint runs through sublime-eslint
```

**What did you expect to happen?**
this is proper use of `this` (the function will be called once it's inside an object
https://eslint.org/docs/rules/no-invalid-this
>The function is on an object literal.
**What actually happened? Please include the actual, raw output from ESLint.**
```
SublimeLinter: #215 SublimeLinter-eslint.sublime-package:60: eslint output:
[{'errorCount': 0,
  'filePath': '<text>',
  'fixableErrorCount': 0,
  'fixableWarningCount': 0,
  'messages': [{'column': 15,
                'endColumn': 19,
                'endLine': 20,
                'line': 20,
                'message': "Unexpected 'this'.",
                'nodeType': 'ThisExpression',
                'ruleId': 'babel/no-invalid-this',
                'severity': 1,
                'source': "            (!this.nextStatus && this.status === 'used') ||"},
               {'column': 34,
                'endColumn': 38,
                'endLine': 20,
                'line': 20,
                'message': "Unexpected 'this'.",
                'nodeType': 'ThisExpression',
                'ruleId': 'babel/no-invalid-this',
                'severity': 1,
                'source': "            (!this.nextStatus && this.status === 'used') ||"},
               {'column': 14,
                'endColumn': 18,
                'endLine': 21,
                'line': 21,
                'message': "Unexpected 'this'.",
                'nodeType': 'ThisExpression',
                'ruleId': 'babel/no-invalid-this',
                'severity': 1,
                'source': "            (this.nextStatus === 'used')"}],
  'source': "export function createProperty (type, value, options = {}) {\n  return {\n    [type]      : value,\n    configurable: true,\n    ...options\n  }\n}\n\nexport function createGetter (value, options) {\n  return createProperty('get', value, options)\n}\n\nexport default function createModelInstance (instance) {\n  return Object.defineProperties(\n    {...instance},\n    {\n      checked: createGetter(\n        function () {\n          return (\n            (!this.nextStatus && this.status === 'used') ||\n            (this.nextStatus === 'used')\n          )\n        }\n      )\n    }\n  )\n}\n",
  'warningCount': 3}]
```

**Comment by platinumazure** [comment]
The function isn't really on an object literal, is it? Looks like the function expression is being passed as an argument to a function call (`createGetter`). The rule can't know that this function will effectively be used on the object.

It seems this is working as designed and is an inherent limitation of `no-invalid-this`. I would recommend using inline configuration comments to disable the lint report for that rule at that location.

**Comment by ThaJay** [comment]
That is true, if it wasn't for `greateGetter` they would actually be on an object literal. I thought of that fact after posting this and understand eslint will not look beyond the first layer.

I decided to keep the rule enabled and add `/*@this*/` like so:

```
createValue(/*@this*/
        function checked () {
```

Thank you for your insight.

**Comment by kaicataldo** [comment]
Closing this issue as it looks like the question has been answered. Please feel free to visit us in the [ESLint Gitter](https://gitter.im/eslint/eslint) if you have any other issues!
